From 02038f4ce933b5d63ea2e4f06a0231cd8f3aa257 Mon Sep 17 00:00:00 2001
From: Yuvarani V <quic_yuvarani@quicinc.com>
Date: Fri, 24 Nov 2023 13:35:37 +0530
Subject: [PATCH] nl80211: Add attribute to pass critical update params

Add NL80211_ATTR_MLD_CRITICAL_UPDATE attribute to pass critical update
params to hostapd through NL80211_CMD_FRAME cmd.

Signed-off-by: Rathees Kumar R Chinannan <quic_rrchinan@quicinc.com>
Signed-off-by: Yuvarani V <quic_yuvarani@quicinc.com>
---
 include/net/cfg80211.h       |   9 +++
 include/uapi/linux/nl80211.h |  70 +++++++++++++++++++
 net/wireless/nl80211.c       | 129 ++++++++++++++++++++++++++++++++++-
 3 files changed, 207 insertions(+), 1 deletion(-)

--- a/include/net/cfg80211.h
+++ b/include/net/cfg80211.h
@@ -6362,7 +6362,11 @@ void wiphy_delayed_work_flush(struct wip
  * @links[].cac_started: true if DFS channel availability check has been started
  * @links[].cac_start_time: timestamp (jiffies) when the dfs state was entered.
  * @links[].cac_time_ms: CAC time in ms
+ * @links[].critical_flag: Critical update bit for the bss
+ * @links[].bpcc: Bss param change count value for the bss
+ * @links[].switch_count: CSA/CCA count for the bss
  * @valid_links: bitmap describing what elements of @links are valid
+ * @critical_update: critical params updated on anyone wdev link
  */
 struct wireless_dev {
 	struct wiphy *wiphy;
@@ -6476,10 +6480,14 @@ struct wireless_dev {
 		bool cac_started;
 		unsigned long cac_start_time;
 		unsigned int cac_time_ms;
+		bool critical_flag;
+		u8 bpcc;
+		u8 switch_count;
 	} links[IEEE80211_MLD_MAX_NUM_LINKS];
 	u16 valid_links;
 
 	u8 reg_6g_power_mode;
+	bool critical_update;
 };
 
 static inline const u8 *wdev_address(struct wireless_dev *wdev)
@@ -8560,6 +8568,7 @@ struct cfg80211_rx_info {
 	u32 flags;
 	u64 rx_tstamp;
 	u64 ack_tstamp;
+	bool critical_update;
 };
 
 /**
--- a/net/wireless/nl80211.c
+++ b/net/wireless/nl80211.c
@@ -485,6 +485,25 @@ nl80211_sta_wme_policy[NL80211_STA_WME_M
 	[NL80211_STA_WME_MAX_SP] = { .type = NLA_U8 },
 };
 
+static const struct nla_policy
+link_policy[NL80211_CU_MLD_LINK_ATTR_MAX + 1] = {
+	[NL80211_CU_MLD_LINK_ATTR_ID] = { .type = NLA_U8 },
+	[NL80211_CU_MLD_LINK_ATTR_CRITICAL_FLAG] = { .type = NLA_FLAG },
+	[NL80211_CU_MLD_LINK_ATTR_BPCC] = { .type = NLA_U8 },
+	[NL80211_CU_MLD_LINK_ATTR_SWITCH_COUNT] = { .type = NLA_U8 },
+};
+
+static const struct nla_policy
+mld_policy[NL80211_CU_MLD_ATTR_MAX + 1] = {
+	[NL80211_CU_MLD_ATTR_IFINDEX] = { .type = NLA_U32 },
+	[NL80211_CU_MLD_ATTR_LINK_LIST] = NLA_POLICY_NESTED(link_policy),
+};
+
+static const struct nla_policy
+cu_policy[NL80211_CU_ATTR_MAX + 1] = {
+	[NL80211_CU_ATTR_MLD_LIST] = NLA_POLICY_NESTED(mld_policy),
+};
+
 static struct netlink_range_validation nl80211_punct_bitmap_range = {
 	.min = 0,
 	.max = 0xffff,
@@ -867,6 +886,7 @@ static const struct nla_policy nl80211_p
 #endif
 	[NL80211_ATTR_ADD_MULTI_CHAN] = { .type = NLA_U32 },
 	[NL80211_ATTR_DEL_MULTI_CHAN] = { .type = NLA_U32 },
+	[NL80211_ATTR_RXMGMT_CRITICAL_UPDATE] = NLA_POLICY_NESTED(cu_policy),
 };
 
 /* policy for the key attributes */
@@ -19816,6 +19836,110 @@ bool cfg80211_rx_unexpected_4addr_frame(
 }
 EXPORT_SYMBOL(cfg80211_rx_unexpected_4addr_frame);
 
+static int nl80211_send_mgmt_critical_update_len(struct wireless_dev *wdev)
+{
+	struct wiphy *wiphy = wdev->wiphy;
+	struct cfg80211_registered_device *rdev = wiphy_to_rdev(wiphy);
+	struct wireless_dev *tmp_wdev;
+	int link_id;
+	int cu_len = 0;
+
+	/*Add length for critical_update and mld_list nla header */
+	cu_len += 12;
+	list_for_each_entry(tmp_wdev, &rdev->wiphy.wdev_list, list) {
+		/* Add length for mld,link list nla header and
+		 * length for WDEV ifindex
+		 */
+		cu_len += 24;
+		for_each_valid_link(tmp_wdev, link_id) {
+			/*Add length for link nla_header and
+			 * length for link_id, critical flag, bpcc and CSA count
+			 */
+			cu_len += 24;
+		}
+	}
+	return cu_len;
+}
+
+static int nl80211_send_mgmt_critical_update(struct sk_buff *msg, struct wireless_dev *wdev)
+{
+	struct wiphy *wiphy = wdev->wiphy;
+	struct wireless_dev *tmp_wdev;
+	struct cfg80211_registered_device *rdev = wiphy_to_rdev(wiphy);
+	struct nlattr *critical_update;
+	struct nlattr *mld_list, *mld;
+	struct nlattr *link_list, *link;
+	struct net_device *tmp_netdev;
+	int link_id;
+	int i = 0, j = 0;
+
+	critical_update = nla_nest_start_noflag(msg,
+						NL80211_ATTR_RXMGMT_CRITICAL_UPDATE);
+	if (!critical_update)
+		goto nla_fail;
+
+	mld_list = nla_nest_start_noflag(msg,
+					 NL80211_CU_ATTR_MLD_LIST);
+	if (!mld_list)
+		goto nla_fail_cu;
+
+	list_for_each_entry(tmp_wdev, &rdev->wiphy.wdev_list, list) {
+		if (!tmp_wdev->valid_links)
+			continue;
+		if (!tmp_wdev->critical_update)
+			continue;
+		mld = nla_nest_start_noflag(msg, i + 1);
+		if (!mld)
+			goto nla_fail_mld_list;
+		tmp_netdev = tmp_wdev->netdev;
+		if (tmp_netdev &&
+		    nla_put_u32(msg, NL80211_CU_MLD_ATTR_IFINDEX, tmp_netdev->ifindex))
+			goto nla_fail_mld;
+		link_list = nla_nest_start_noflag(msg,
+						  NL80211_CU_MLD_ATTR_LINK_LIST);
+		if (!link_list)
+			goto nla_fail_mld;
+
+		for_each_valid_link(tmp_wdev, link_id) {
+			link = nla_nest_start_noflag(msg, j + 1);
+			if (!link)
+				goto nla_fail_link_list;
+			if (nla_put_u8(msg, NL80211_CU_MLD_LINK_ATTR_ID, link_id))
+				goto nla_fail_link;
+			if (tmp_wdev->links[link_id].critical_flag &&
+			    nla_put_flag(msg, NL80211_CU_MLD_LINK_ATTR_CRITICAL_FLAG))
+				goto nla_fail_link;
+			if (nla_put_u8(msg, NL80211_CU_MLD_LINK_ATTR_BPCC,
+				       tmp_wdev->links[link_id].bpcc))
+				goto nla_fail_link;
+			if (nla_put_u8(msg, NL80211_CU_MLD_LINK_ATTR_SWITCH_COUNT,
+				       tmp_wdev->links[link_id].switch_count))
+				goto nla_fail_link;
+			nla_nest_end(msg, link);
+			j++;
+		}
+		nla_nest_end(msg, link_list);
+		nla_nest_end(msg, mld);
+		i++;
+	}
+	nla_nest_end(msg, mld_list);
+	nla_nest_end(msg, critical_update);
+	return 0;
+
+nla_fail_link:
+	nla_nest_cancel(msg, link);
+nla_fail_link_list:
+	nla_nest_cancel(msg, link_list);
+nla_fail_mld:
+	nla_nest_cancel(msg, mld);
+nla_fail_mld_list:
+	nla_nest_cancel(msg, mld_list);
+nla_fail_cu:
+	nla_nest_cancel(msg, critical_update);
+nla_fail:
+	return -ENOBUFS;
+}
+
 int nl80211_send_mgmt(struct cfg80211_registered_device *rdev,
 		      struct wireless_dev *wdev, u32 nlportid,
 		      struct cfg80211_rx_info *info, gfp_t gfp)
@@ -19823,8 +19947,12 @@ int nl80211_send_mgmt(struct cfg80211_re
 	struct net_device *netdev = wdev->netdev;
 	struct sk_buff *msg;
 	void *hdr;
+	int cu_len = 0;
+
+	if (info->critical_update)
+		cu_len = nl80211_send_mgmt_critical_update_len(wdev);
 
-	msg = nlmsg_new(100 + info->len, gfp);
+	msg = nlmsg_new(100 + info->len + cu_len, gfp);
 	if (!msg)
 		return -ENOMEM;
 
@@ -19858,6 +19986,11 @@ int nl80211_send_mgmt(struct cfg80211_re
 						   NL80211_ATTR_PAD)))
 		goto nla_put_failure;
 
+	if (info->critical_update) {
+		if (nl80211_send_mgmt_critical_update(msg, wdev))
+			goto nla_put_failure;
+		wdev->critical_update = 0;
+	}
 	genlmsg_end(msg, hdr);
 
 	return genlmsg_unicast(wiphy_net(&rdev->wiphy), msg, nlportid);
